staging/src/k8s.io/apiserver/pkg/util/flowcontrol/reqmgmt-config.go

```diff
@@ -0,0 +1,350 @@
+	var shareSum float64
+	newRMState := &requestManagerState{
+		priorityLevelStates: make(map[string]*priorityLevelState),
+	}
+	newlyQuiescent := make([]*priorityLevelState, 0)
+	var haveExemptPL, haveCatchAllPL, haveExemptFS, haveCatchAllFS bool
+	for _, pl := range newPLs {
+		state := oldRMState.priorityLevelStates[pl.Name]
+		if state == nil {
+			state = &priorityLevelState{
+				config: pl.Spec,
+			}
+		} else {
+			oState := *state
+			state = &oState
+			state.config = pl.Spec
+			if state.emptyHandler != nil { // it was undesired, but no longer
+				klog.V(3).Infof("Priority level %q was undesired and has become desired again", pl.Name)
+				state.emptyHandler = nil
+				state.queues.Quiesce(nil)
+			}
+		}
+		if state.config.Limited != nil {
+			shareSum += float64(state.config.Limited.AssuredConcurrencyShares)
+			qsConfig, err := qscOfPL(pl, reqMgr.requestWaitLimit)
+			if err != nil {
+				klog.Warningf(err.Error())
+				continue
+			}
+			state.qsConfig = qsConfig
+		} else {
+			haveExemptPL = true
+		}
+		haveCatchAllPL = haveCatchAllPL || pl.Name == rmtypesv1a1.PriorityLevelConfigurationNameCatchAll
+		newRMState.priorityLevelStates[pl.Name] = state
+	}
+
+	fsSeq := make(apihelpers.FlowSchemaSequence, 0, len(newFSs))
+	for i, fs := range newFSs {
+		_, flowSchemaExists := newRMState.priorityLevelStates[fs.Spec.PriorityLevelConfiguration.Name]
+		reqMgr.syncFlowSchemaStatus(fs, !flowSchemaExists)
+		if flowSchemaExists {
+			fsSeq = append(fsSeq, newFSs[i])
+			haveExemptFS = haveExemptFS || fs.Name == rmtypesv1a1.FlowSchemaNameExempt
+			haveCatchAllFS = haveCatchAllFS || fs.Name == rmtypesv1a1.FlowSchemaNameCatchAll
+		}
+	}
+	sort.Sort(fsSeq)
+
+	for plName, plState := range oldRMState.priorityLevelStates {
+		if newRMState.priorityLevelStates[plName] != nil {
+			// Still desired and already updated
+			continue
+		}
+		newState := *plState
+		plState = &newState
+		if plState.emptyHandler != nil && plState.emptyHandler.IsEmpty() {
+			// The queues are empty and will never get any more requests
+			plState.queues = nil
+			plState.emptyHandler = nil
+			klog.V(3).Infof("Retired queues for undesired quiescing priority level %q", plName)
+		}
+		if plState.config.Limited == nil && !haveExemptPL || plName == rmtypesv1a1.PriorityLevelConfigurationNameCatchAll && !haveCatchAllPL {
+			klog.V(3).Infof("Retaining old priority level %q with Type=%v because of lack of replacement", plName, plState.config.Type)
+		} else {
+			if plState.queues == nil { // should never happen; but if it does:
+				klog.V(3).Infof("Immediately removing undesired priority level %q, Type=%v", plName, plState.config.Type)
+				continue
+			}
+			if plState.emptyHandler == nil {
+				klog.V(3).Infof("Priority level %q became undesired", plName)
+				plState.emptyHandler = &emptyRelay{reqMgr: reqMgr}
+				newlyQuiescent = append(newlyQuiescent, plState)
+			}
+		}
+		if plState.config.Limited != nil {
```

Comment:
if we only check if config.Limited is not nil, it means that acs of every old non-exempt priority will be taken into consideration. Is it expected? I think for those "abandoned" old plstate, only acs of quiescent ones should be calculated. Shall we add plState.emptyHandler != nil to the if statement?

Reply:
Yes, we want every retained non-exempt priority level to participate in the divvying up of the overall concurrency limit.  This is so that the queues of all those priority levels can continue to drain.